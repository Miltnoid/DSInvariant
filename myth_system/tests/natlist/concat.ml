type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type llist =
  | LNil
  | LCons of list * llist

(*
let append : list -> list -> list =
  let rec g (xs1:list) (xs2:list) : list =
    match xs1 with
    | Nil -> xs2
    | Cons (x, xs1p) -> Cons (x, g xs1p xs2)
  in
    g
;;
*)

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (hd, tl) -> Cons (hd, append tl l2)
;;

let concat : llist -> list |>
  { LNil => []
  | LCons ([], LNil) => []
  | LCons ([0], LNil) => [0]
  | LCons ([0], LCons([0], LNil)) => [0;0]
  | LCons ([1], LNil) => [1]
  | LCons ([1], LCons([1], LNil)) => [1;1]
  } = ?
