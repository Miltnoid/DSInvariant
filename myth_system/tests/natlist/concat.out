(* Generated Wed Dec 31 11:16:14 2014 *)
let concat : llist -> list =
  let rec f1 (l1:llist) : list =
    match l1 with
      | LNil -> Nil
      | LCons (l2, l3) -> append (f1 l3) l2
  in
    f1
;;
